### PR TITLE
Implement remove_tokens_for_client()

### LINKED ISSUE #640: Feature Request: Removing app tokens
Problem description: Historically, [there is no api in MSAL to remove tokens obtained from `AcquireTokenForClient()`](https://learn.microsoft.com/en-us/entra/msal/dotnet/acquiring-tokens/clear-token-cache#web-api-and-daemon-apps). Therefore, there is no way for an app to "log out" a service principal.

Proposal: MSALs add a new API ~`ConfidentialClientApplication.RemoveAppTokens()`~ `ConfidentialClientApplication.remove_tokens_for_client()`.

* It will purge app tokens that match the current app's authority (host + tenant) and client_id.
* It does not accept a `scope` parameter, because our intention is to purge all app tokens for the current app, regardless of their scopes.
* It will NOT purge user tokens obtained by this app.

### LINKED ISSUE #650: [Feature Request] Support `force_refresh` for service principal
### MSAL client type

Confidential

### Problem Statement

See https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/649 for details.

### Proposed solution

It should be possible to `force_refresh` access tokens for service principals.

### PR DESCRIPTION
**Pull request by rayluo**
Based on [this understanding](https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/650#issuecomment-1935768909), here comes this ([evil?](https://en.wikipedia.org/wiki/Number_of_the_beast)) PR, which will resolve #640 and resolve #650

### CODE REVIEW

**Review by bgavrilMS — APPROVED** [review]

File: msal/application.py [hunk 1]
```diff
@@ -2178,6 +2178,19 @@ def _acquire_token_for_client(
         telemetry_context.update_telemetry(response)
         return response
 
+    def remove_tokens_for_client(self):
+        """Remove all tokens that were previously acquired via
```
[the remark this answers is not in the document]
Comment (rayluo) [reply msal/application.py:2182]: Good question. The intention is that removing a CCA's token shall leave other MSI's tokens intact, and vice versa. I'll expand the test case when I polish the Managed Identity feature in the near future.

By the way, has MSAL .Net also (plan to or already?) implemented CCA's AcquireTokenForClient() for Managed Identity scenario? I believe I read similar description in SLC communications.

**Comment by jiasli** [comment]
> here comes this ([evil?](https://en.wikipedia.org/wiki/Number_of_the_beast)) PR

Interestingly, 666 is a positive/luckily number in Chinese, expressing "good" and "proficient": https://en.wikipedia.org/wiki/666_(number).

**Comment by yonzhan** [comment]
[celebrate]     Yong Zhang reacted to your message:
________________________________

**Comment by jiasli** [comment]
The implementation is not complete. After running `remove_tokens_for_client()`, the ghost of the service principal still lingers in `AppMetadata`:

```
{
    "AccessToken": {},
    "AppMetadata": {
        "appmetadata-login.microsoftonline.com-a7003d8c-e50f-4371-91a6-ef37bba4ab23": {
            "client_id": "a7003d8c-e50f-4371-91a6-ef37bba4ab23",
            "environment": "login.microsoftonline.com"
        }
    }
}
```

File: msal/application.py [hunk 2]
```diff
@@ -2178,6 +2178,19 @@ def _acquire_token_for_client(
...
+        :func:`~acquire_token_for_client()` for the current client."""
+        for env in [self.authority.instance] + self._get_authority_aliases(
+                self.authority.instance):
+            for at in self.token_cache.find(TokenCache.CredentialType.ACCESS_TOKEN, query={
+                "client_id": self.client_id,
+                "environment": env,
+                "home_account_id": None,  # These are mostly app-only tokens
+            }):
```
Comment (jiasli) [inline msal/application.py:2190]: Luckily `realm` (tenant) is not used in `query`. This makes the functionality align with `az logout` which only takes a `--username`, but not `--tenant`.

This means all access tokens for a service principal from all tenants will be removed. However, this brings up another question: what if I only want to remove the access token for a service principal in a specific tenant?

File: msal/application.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jiasli) [reply msal/application.py:2182]: I created https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/726 to track `remove_tokens_for_client()` support in `ManagedIdentityClient`.

**Comment by jiasli** [comment]
What if I lose track of service principals client IDs? How do I purge all service principals' access tokens, while keeping users' access tokens?

**Comment by rayluo** [comment]
> What if I lose track of service principals client IDs? How do I purge all service principals' access tokens, ...?

Hmm, MSAL's `ClientApplication` always requires `client_id` as its required parameter. This question sounds as strange as "what if I forgot my local Windows username, how do I purge all users' content from the current PC?". Is there a legit scenario for that? That being said, I suppose you can do `format c:` or its equivalent `rm ~/.azure/msal_token_cache.json` if desirable.

> How do I purge all service principals' access tokens, while keeping users' access tokens?

You would have to create different `ClientApplication` objects with different token cache persistence files, so that you may be able to delete some of those files.

**Comment by rayluo** [comment]
> The implementation is not complete. After running `remove_tokens_for_client()`, the ghost of the service principal still lingers in `AppMetadata`:
> 
> ```
> {
>     "AccessToken": {},
>     "AppMetadata": {
>         "appmetadata-login.microsoftonline.com-a7003d8c-e50f-4371-91a6-ef37bba4ab23": {
>             "client_id": "a7003d8c-e50f-4371-91a6-ef37bba4ab23",
>             "environment": "login.microsoftonline.com"
>         }
>     }
> }
> ```

The residue of `AppMetadata` is technically not an incomplete implementation. The `AppMetadata` was meant to be persisted forever, because (1) they would be used to remember whether an app belongs to a "family" (although that characteristic does not necessarily apply to service principal); (2) they contain no sensitive tokens anyway.

**Comment by jiasli** [comment]
> This question sounds as strange as "what if I forgot my local Windows username, how do I purge all users' content from the current PC?".

This is possible for user accounts logged into MSAL as users can be retrieved with `ClientApplication.get_accounts()`:

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/3279f045dc573d8408ddcdf3565e8b16247627be/msal/application.py#L1162

Azure CLI has a [`logout_all_users()`](https://github.com/Azure/azure-cli/blob/461d7b0573885f9426769ab73e2f1e4c026ee30b/src/azure-cli-core/azure/cli/core/auth/identity.py#L212-L220) which removes all users, but [`logout_service_principal()`](https://github.com/Azure/azure-cli/blob/461d7b0573885f9426769ab73e2f1e4c026ee30b/src/azure-cli-core/azure/cli/core/auth/identity.py#L233-L238) is currently incomplete due to no equivalent of `ClientApplication.get_accounts()`.

**Comment by jiasli** [comment]
> although that characteristic does not necessarily apply to service principal

I agree the client ID for user authentication (such as Azure CLI's client ID) is not sensitive, but some users may treat service principal's client ID as sensitive date and want it to be "obliviated".

**Pull request merged**